lib/improver/utilities/cube_units.py

```diff
@@ -32,14 +32,125 @@
+    """
+    Function to check the units and datatypes of cube diagnostics and
+    coordinates against the manifest in improver.units, with option to
+    enforce or fail for non-conforming data.
+
+    Args:
+        cubes (iris.cube.Cube or iris.cube.CubeList):
+            Cube or list of cubes to be checked
+
+    Kwargs:
+        coords (list or None):
+            List of coordinate names to check.  If None, checks all
+            coordinates present on the input cubes.
+        enforce (bool):
+            If True, this function returns a list of conformant cubes.
+            If False, a ValueError is thrown if the cubes do not conform.
+
+    Raises:
+        ValueError: if "enforce=False" and the input cubes do not conform
+            to the datatypes and units standard.
+
+    Returns:
+        new_cubes (iris.cube.CubeList):
+            New cubelist with conformant datatypes and units
+    """
+    # convert input to CubeList
+    if isinstance(cubes, iris.cube.Cube):
+        cubes = [cubes]
+
+    # make a copy of the input cube list
+    new_cubes = enforce_diagnostic_units_and_dtypes(cubes, inplace=False)
+    # set up coordinates to check
+    if coords is not None:
+        all_coords = list(set(coords))
+    else:
+        all_coords = []
+        for cube in cubes:
+            all_coords.extend(
+                [coord.name() for coord in new_cubes[0].coords()])
+        all_coords = list(set(all_coords))
+    # modify the copied cubes in place
+    enforce_coordinate_units_and_dtypes(new_cubes, all_coords)
+
+    if not enforce:
+        # check each cube against its counterpart and fail if changed
+        for cube, ref in zip(new_cubes, cubes):
+            if cube.units != ref.units:
+                msg = ('Units {} of {} cube do not conform to '
+                       'expected standard {}')
+                raise ValueError(msg.format(ref.units, ref.name(), cube.units))
+            if cube.dtype != ref.dtype:
+                msg = ('Datatype {} of {} cube does not conform to '
+                       'expected standard {}')
+                raise ValueError(msg.format(ref.dtype, ref.name(), cube.dtype))
+            for coord in all_coords:
+                try:
+                    if cube.coord(coord).units != ref.coord(coord).units:
+                        msg = ('Units {} of coordinate {} on {} cube do not '
+                               'conform to expected standard')
+                        raise ValueError(msg.format(
+                            cube.coord(coord).units, ref.coord(coord).name(),
+                            ref.name()))
+                    if cube.coord(coord).dtype != ref.coord(coord).dtype:
+                        msg = ('Datatype {} of coordinate {} on {} cube does '
+                               'not conform to expected standard')
+                        raise ValueError(msg.format(
+                            cube.coord(coord).dtype, ref.coord(coord).name(),
```

Comment:
I think this should actually be `ref.coord(coord).dtype`, rather than `cube.coord(coord).dtype`. What do you think?

Reply:
Yep!